distsqlrun: short circuit hash joins on empty side

rjnn opened the pull request:
Previously, hash joins would consume both sides in their entirety,
even if one side was empty. In the case of inner joins, this is
unnecessary, since there can be no matches. Short circuit the
hashjoiner if one side is empty, and return early. Fixes #18841.

Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/26485)
<!-- Reviewable:end -->

jordanlewis commented:
Thanks! FYI I don't think the Fixes declaration will work because it's not on its own line.

---

Review status: 0 of 3 files reviewed at latest revision, all discussions resolved, some commit checks pending.

---

*[pkg/sql/distsqlrun/hash_row_container.go, line 86 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26485#-LELrhUYDeD40i7pDtgH:-LELrhUYDeD40i7pDtgI:b-9u80j7) ([raw file](https://github.com/cockroachdb/cockroach/blob/501b045940605990aa2acfc7d24ab221a3c31693/pkg/sql/distsqlrun/hash_row_container.go#L86)):*
> ```Go
> 	// Close frees up resources held by the hashRowContainer.
> 	Close(context.Context)
> 	// Empty indicates that there are zero rows in the hashRowContainer
> ```

take-it-or-leave-it: this is a method, not a field, so I'd phrase the godoc as "Empty returns true if..."

---

*[pkg/sql/distsqlrun/hash_row_container.go, line 209 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26485#-LELsA4zDoRnKAC5SB6F:-LELsA4zDoRnKAC5SB6G:b7vhl34) ([raw file](https://github.com/cockroachdb/cockroach/blob/501b045940605990aa2acfc7d24ab221a3c31693/pkg/sql/distsqlrun/hash_row_container.go#L209)):*
> ```Go
> 
> func (h *hashMemRowContainer) Empty() bool {
> 	return h.memRowContainer.Len() == 0
> ```

I was going to say you should add quick unit tests for these, and then I realized that this struct doesn't have any tests at all. Did I just miss them? Anyway, could you please add some tests for at least this method?

In general it would be good to test this very important infrastructural piece, and benchmark it, obviously you don't have to do all that in this PR though.

---

*[pkg/sql/distsqlrun/hashjoiner.go, line 303 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26485#-LELsg7N2KXm4scvEpkA:-LELsg7O89EE1e9cI_ua:b-8w07ie) ([raw file](https://github.com/cockroachdb/cockroach/blob/501b045940605990aa2acfc7d24ab221a3c31693/pkg/sql/distsqlrun/hashjoiner.go#L303)):*
> ```Go
> 	// If storedSide is empty and we are processing an inner-join, we can early exit.
> 	if h.joinType == sqlbase.InnerJoin && storedRows.Empty() {
> 		DrainAndClose(ctx, h.out.output, err /* cause */, h.pushTrailingMeta, h.leftSource, h.rightSource)
> ```

This looks fine, but I think we can do it earlier, before we even try to build the hash container. I understand that it won't matter, since a 0-sized container will be built instantly, but nevertheless I think it would be clearer to do this directly after `bufferPhase` instead of after `buildPhase`.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26485#-:-LELrXDt48cR2I3lGTuz:brtmt9g)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
> FYI I don't think the Fixes declaration will work because it's not on its
own line.

I think it will according to this:
https://help.github.com/articles/closing-issues-using-keywords/

On Wed, Jun 6, 2018 at 5:02 PM, Jordan Lewis <notifications@github.com>
wrote:

> Thanks! FYI I don't think the Fixes declaration will work because it's not
> on its own line.
> ------------------------------
>
> Review status: 0 of 3 files reviewed at latest revision, all discussions
> resolved, some commit checks pending.
> ------------------------------
>
> *pkg/sql/distsqlrun/hash_row_container.go, line 86 at r1
> <https://reviewable.io/reviews/cockroachdb/cockroach/26485#-LELrhUYDeD40i7pDtgH:-LELrhUYDeD40i7pDtgI:b-9u80j7>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/501b045940605990aa2acfc7d24ab221a3c31693/pkg/sql/distsqlrun/hash_row_container.go#L86>):*
>
> 	// Close frees up resources held by the hashRowContainer.
> 	Close(context.Context)
> 	// Empty indicates that there are zero rows in the hashRowContainer
>
> take-it-or-leave-it: this is a method, not a field, so I'd phrase the
> godoc as "Empty returns true if..."
> ------------------------------
>
> *pkg/sql/distsqlrun/hash_row_container.go, line 209 at r1
> <https://reviewable.io/reviews/cockroachdb/cockroach/26485#-LELsA4zDoRnKAC5SB6F:-LELsA4zDoRnKAC5SB6G:b7vhl34>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/501b045940605990aa2acfc7d24ab221a3c31693/pkg/sql/distsqlrun/hash_row_container.go#L209>):*
>
> func (h *hashMemRowContainer) Empty() bool {
> 	return h.memRowContainer.Len() == 0
>
> I was going to say you should add quick unit tests for these, and then I
> realized that this struct doesn't have any tests at all. Did I just miss
> them? Anyway, could you please add some tests for at least this method?
>
> In general it would be good to test this very important infrastructural
> piece, and benchmark it, obviously you don't have to do all that in this PR
> though.
> ------------------------------
>
> *pkg/sql/distsqlrun/hashjoiner.go, line 303 at r1
> <https://reviewable.io/reviews/cockroachdb/cockroach/26485#-LELsg7N2KXm4scvEpkA:-LELsg7O89EE1e9cI_ua:b-8w07ie>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/501b045940605990aa2acfc7d24ab221a3c31693/pkg/sql/distsqlrun/hashjoiner.go#L303>):*
>
> 	// If storedSide is empty and we are processing an inner-join, we can early exit.
> 	if h.joinType == sqlbase.InnerJoin && storedRows.Empty() {
> 		DrainAndClose(ctx, h.out.output, err /* cause */, h.pushTrailingMeta, h.leftSource, h.rightSource)
>
> This looks fine, but I think we can do it earlier, before we even try to
> build the hash container. I understand that it won't matter, since a
> 0-sized container will be built instantly, but nevertheless I think it
> would be clearer to do this directly after bufferPhase instead of after
> buildPhase.
> ------------------------------
>
> *Comments from Reviewable
> <https://reviewable.io/reviews/cockroachdb/cockroach/26485#-:-LELrXDt48cR2I3lGTuz:brtmt9g>*
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/26485#issuecomment-395212767>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAXBcdD_l1A5XvjbV2vVhrVFvZfCD-4lks5t6ENzgaJpZM4UdWL1>
> .
>

jordanlewis commented:
TIL!

jordanlewis commented:
![image](https://user-images.githubusercontent.com/43821/41065336-5913bc72-69ac-11e8-8a90-32d097805d43.png)
The UI even says so, how about that!

RaduBerinde commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 3 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/sql/distsqlrun/hashjoiner.go, line 301 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26485#-LEMvc6R4nTdTRpIrYeC:-LEMvc6SBA2A4uKI45vd:b-a9lyhl) ([raw file](https://github.com/cockroachdb/cockroach/blob/501b045940605990aa2acfc7d24ab221a3c31693/pkg/sql/distsqlrun/hashjoiner.go#L301)):*
> ```Go
> 	log.VEventf(ctx, 1, "build phase complete")
> 
> 	// If storedSide is empty and we are processing an inner-join, we can early exit.
> ```

Can also exit for a left/right join, depending on the side

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26485#-:-LEMvv308N94AoY1h-y-:bnfp4nl)*
<!-- Sent from Reviewable.io -->

asubiotto commented:
Review status: 0 of 3 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/sql/distsqlrun/hashjoiner.go, line 303 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26485#-LELsg7N2KXm4scvEpkA:-LEPaPWB3ly-glOFcuHq:b-tkwchs) ([raw file](https://github.com/cockroachdb/cockroach/blob/501b045940605990aa2acfc7d24ab221a3c31693/pkg/sql/distsqlrun/hashjoiner.go#L303)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

This looks fine, but I think we can do it earlier, before we even try to build the hash container. I understand that it won't matter, since a 0-sized container will be built instantly, but nevertheless I think it would be clearer to do this directly after `bufferPhase` instead of after `buildPhase`.
</blockquote></details>

+1 I would even recommend doing it in the `bufferPhase` right around where the `// This stream is done, great!` comment is. You can just return `nil, true, nil` to indicate that we should exit. You'll also be using a `memRowContainer` at that point so you can call its `Len` method directly and remove the code to the `hashRowContainer`s

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26485)*
<!-- Sent from Reviewable.io -->

Commit:
distsqlrun: short circuit hash joins on empty side

Previously, hash joins would consume both sides in their entirety,
even if one side was empty. In the case of inner joins, this is
unnecessary, since there can be no matches. Short circuit the
hashjoiner if one side is empty, and return early. Fixes #18841.

Release note: None

arjunravinarayan commented:
Review status: 0 of 1 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[pkg/sql/distsqlrun/hashjoiner.go, line 301 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26485#-LEMvc6R4nTdTRpIrYeC:-LEQEROb8Mn6At6Gqsn4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/501b045940605990aa2acfc7d24ab221a3c31693/pkg/sql/distsqlrun/hashjoiner.go#L301)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Can also exit for a left/right join, depending on the side
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/hashjoiner.go, line 303 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26485#-LELsg7N2KXm4scvEpkA:-LEQENjRE8WzrvTTfU89:b-rhdlj9) ([raw file](https://github.com/cockroachdb/cockroach/blob/501b045940605990aa2acfc7d24ab221a3c31693/pkg/sql/distsqlrun/hashjoiner.go#L303)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

+1 I would even recommend doing it in the `bufferPhase` right around where the `// This stream is done, great!` comment is. You can just return `nil, true, nil` to indicate that we should exit. You'll also be using a `memRowContainer` at that point so you can call its `Len` method directly and remove the code to the `hashRowContainer`s
</blockquote></details>

Done. Great tips @asubiotto! Much appreciated :raised_hands: 

---

*[pkg/sql/distsqlrun/hash_row_container.go, line 86 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26485#-LELrhUYDeD40i7pDtgH:-LEPvOCIBKDbjoOAnx8p:b-qth5cd) ([raw file](https://github.com/cockroachdb/cockroach/blob/501b045940605990aa2acfc7d24ab221a3c31693/pkg/sql/distsqlrun/hash_row_container.go#L86)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

take-it-or-leave-it: this is a method, not a field, so I'd phrase the godoc as "Empty returns true if..."
</blockquote></details>

Tooked.

---

*[pkg/sql/distsqlrun/hash_row_container.go, line 209 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26485#-LELsA4zDoRnKAC5SB6F:-LEPvZKbDQ_Rx7MM3w2f:bl7hy6c) ([raw file](https://github.com/cockroachdb/cockroach/blob/501b045940605990aa2acfc7d24ab221a3c31693/pkg/sql/distsqlrun/hash_row_container.go#L209)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

I was going to say you should add quick unit tests for these, and then I realized that this struct doesn't have any tests at all. Did I just miss them? Anyway, could you please add some tests for at least this method?

In general it would be good to test this very important infrastructural piece, and benchmark it, obviously you don't have to do all that in this PR though.
</blockquote></details>

I'll do a follow up PR. Filed issue #26513 to track the work.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26485)*
<!-- Sent from Reviewable.io -->

arjunravinarayan commented:
*[pkg/sql/distsqlrun/hash_row_container.go, line 86 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/26485#-LELrhUYDeD40i7pDtgH:-LEQJDyn5WeZmYMuitlt:b-w61srk) ([raw file](https://github.com/cockroachdb/cockroach/blob/501b045940605990aa2acfc7d24ab221a3c31693/pkg/sql/distsqlrun/hash_row_container.go#L86)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary><blockquote>

Tooked.
</blockquote></details>

Oh hmm, this was made redundant by @asubiotto's refactor which eliminated this need, so I guess not actually tooked. But I'll keep your method vs field grammatical "tips" in mind for the future.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26485)*
<!-- Sent from Reviewable.io -->

arjunravinarayan commented:
bors r+

---







*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26485#-:-LEQQU-C4oPshnS2m2YV:b-ydi9ay)*
<!-- Sent from Reviewable.io -->

asubiotto reviewed (approved):
LGTM. Probably worth some minor regression test/verification in hashjoiner_test.go if possible. You might be able to use stats here (take a look at https://github.com/cockroachdb/cockroach/blob/ebc9b0d4dac3caf323a3e15729d7b307c47ae5f0/pkg/sql/distsqlrun/tablereader_test.go#L341) to get the number of rows read from the non-zero side. I'm guessing it would have to be a `bufferPhase` unit-test that checks the value of `earlyExit`.

arjunravinarayan commented:
Good point. All my testing was just manually testing it on a local 1-node cluster.
